tests/pytorch/distributed/run_numerics.py

```diff
@@ -0,0 +1,701 @@
+fp8_recipe = DelayedScaling(fp8_format=fp8_format, amax_history_len=32, amax_compute_algo="max")
+
+
+def main(argv=None, namespace=None):
+    global WORLD_RANK, WORLD_SIZE, NCCL_WORLD, FP8
+
+    dist.init_process_group(backend="nccl")
+    NCCL_WORLD = dist.new_group(backend="nccl")
+
+    WORLD_RANK = dist.get_rank()
+    WORLD_SIZE = dist.get_world_size()
+    assert WORLD_SIZE == 2, "This test uses 2 GPUs. Run with torchrun --nproc_per_node=2."
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-l", "--layer-type", type=str)
+    parser.add_argument("--fp8", action="store_true", default=False)
+    args = parser.parse_args(argv, namespace)
+
+    test_dict = {
+        "linear": test_linear,
+        "layernorm": test_layernorm,
+        "layernorm_linear": test_layernorm_linear,
+        "layernorm_mlp": test_layernorm_mlp,
+        "transformer_layer": test_transformer_layer,
+    }
+
+    assert args.layer_type in test_dict, f"Argument --layer-type must be in {test_dict.keys()}"
+
+    FP8 = args.fp8
+
+    output_value = test_dict[args.layer_type]()
+    dist.destroy_process_group()
+    return output_value
+
+
+def run_distributed_test(test_name=None):
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            name = test_name if test_name is not None else func.__name__
+
+            dist_print(f"Starting test {name} with args {args} and {kwargs}")
+            torch.cuda.set_device(WORLD_RANK)
+            torch.manual_seed(1234)
+            torch.cuda.manual_seed(1234)
+            func(*args, **kwargs)
+
+            dist.barrier()
+            dist_print(f"Passed test {name}")
+
+        return wrapper
+
+    return decorator
+
+
+def _gather(tensor, dim=0):
+    """
+    Gathers tensors and concats them. Since torch.distributed.nn.functional.all_gather
+    multiplies gradients by WORLD_SIZE, those gradiedts are rescaled.
+    """
+
+    class HalfGradient(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, input):
+            ctx.save_for_backward(input)
+            return input  # forward pass (identity)
+
+        @staticmethod
+        def backward(ctx, grad_output):
+            return grad_output / WORLD_SIZE  # gradient division by WORLD_SIZE
+
+    tensor = HalfGradient.apply(tensor)
+    gathered = torch.distributed.nn.functional.all_gather(tensor, group=NCCL_WORLD)
+    return torch.cat(gathered, dim=dim)
+
+
+def _constant(tensor):
+    return nn.init.constant_(tensor, 0.5)
+
+
+def dist_print(msg, src=None, end="\n", error=False):
+    stream = sys.stderr if error else sys.stdout
+    if WORLD_RANK == (0 if src is None else src):
+        stream.write(f"[rank{WORLD_RANK}] {msg}{end}\n")
+    dist.barrier()
+
+
+def _check_outputs(output_single_node, output_distributed):
+    numerics_failed = torch.tensor([0], dtype=torch.uint8, device="cuda")
+    rtol = 0.125 if FP8 else 0.025
+    atol = 0.0625 if FP8 else 0.00125
```

Comment:
Ideally we would use the tightest tolerances possible for the dtype, like in [`torch.testing.assert_close`](https://pytorch.org/docs/stable/testing.html#torch.testing.assert_close). I see that the tensor dimensions are small (~64), so we should be able to get away with this.

Reply:
I use the same values as test_comm_gemm_overlap. Tolerances you suggested are not sufficient enough.

Reply:
What layers are you seeing these numerical errors? The reason we have such loose tolerances in the other tests is because the tensor sizes are so large. For the record I think this was a mistake, and it gives us much more confidence if we have tight tolerances with small tensor sizes. We also test many of the same kernels with tight tolerances in [`test_fusible_ops.py`](https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/test_fusible_ops.py) and [`test_fusible_ops_distributed.py`](https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/test_fusible_ops_distributed.py), so I wouldn't expect us to have significant correctness issues.

Reply:
For FP8 layer norm_linear, layernorm_mlp and transformer_layer cast some values to FP8, so I guess that is the reason. I obtain different outputs in Linear with row parallelism. It's behavior of FP8 gemm, which I do not fully understand. It's not because of communication - sum of outputs of these gemms does not equal output of the gemm in non-distributed case by margin 10^-4.

For FP32/FP16 I made all tensor smaller and it seems to be working. I was still getting problem with transformer_layer with layer_type=encoder, so I removed it from the tests.

Reply:
Makes sense, if there's FP8 compute then that is the limiting factor for numerical accuracy. Having tight tolerances for the non-FP8 case is the main thing.